specreduce/background.py

```diff
@@ -85,14 +85,18 @@ def _to_trace(trace):
                     raise ValueError('trace_object.trace_pos must be >= 1')
             return trace
 
+        if self.width < 0:
+            raise ValueError("width must be positive")
+
+        if self.width == 0:
+            self.bkg_array = np.zeros(self.image.shape[self.disp_axis])
+            return
+
         bkg_wimage = np.zeros_like(self.image, dtype=np.float64)
         for trace in self.traces:
             trace = _to_trace(trace)
-            if (np.any(trace.trace.data >= self.image.shape[self.crossdisp_axis]) or
-                    np.any(trace.trace.data < 0)):
-                raise ValueError("center of background window goes beyond image boundaries")
-            elif (np.any(trace.trace.data + self.width/2. >= self.image.shape[self.crossdisp_axis])
-                  or np.any(trace.trace.data - self.width/2. < 0)):
+            if (np.any(trace.trace.data + self.width/2. >= self.image.shape[self.crossdisp_axis])
+                    or np.any(trace.trace.data - self.width/2. < 0)):
                 warnings.warn("background window extends beyond image boundaries")
```

Comment:
Hmmm, that's a good catch, but I'm not sure the best solution (or honestly what would be expected - does this burden fall on the user to set their warning options?  When using interactively in jdaviz, would constant warnings be too loud?).  @pllim - any suggestions or standards we should be following here?

Reply:
Have you tried the "always" setting? https://docs.python.org/3/library/warnings.html#the-warnings-filter

Reply:
I think that's what @ojustino was suggesting, but is that something the user should do or we should override in the code?  I see cases where one warning could easily get lost, but also cases where constant warnings would be too noisy.

Reply:
Ah, sorry, I misunderstood the question. 

Now that I look at the rest of the code, is this warning even that useful? It does not tell you which trace and which image. What do you want user to do about the warning?

If you want it to be very verbose, perhaps a logger is what you need but I think that is a package-wide change of behavior and might be too much work for too little gain.

Another thing you can do is delay the warning. Store them in a variable when you run the loop, and only emit them all at once as a warning at the end of the processing.

Again, a warning is a sign to the user that they should do something different. You should tell them what you want them to change. Otherwise, they will just ignore it if it is too cryptic and unactionable.

Reply:
This warning means that the entirety of the window is out of bounds somewhere along the dispersion axis.  So the user needs to change any one or all of the input window(s).  We could state the values along the dispersion axis (pixel number or wavelength) where this occurs, but I don't see that as being useful information that is actionable.  We could provide advice to lower the separation and/or width, but I don't think those are always the correct thing to do.

Reply:
Would it be more complicated than using `enumerate` in the loop to give the index of the offending trace (e.g., "background window of trace 0 extends...")? That seems more actionable. As for images, this can only apply to one (`self.image`) unless we're thinking about creating several `Background` objects in a loop.

Maybe the warning could end with something roughly like "...if it's important to consider the entire window, consider changing the trace or lowering separation/width." They don't have to do these things, but we at least give examples of strategies that would avoid the warning.

@kecnry, adding an argument or attribute for warning frequency seems like the best solution. I see what you mean for Jdaviz, but someone using strategies to "fix" the regions like I listed above should be able to see feedback on whether the changes worked as intended.

Reply:
I don't think an integer is terribly helpful, since the user either provided a trace and separation or individual traces.  But I tried to make the warning more verbose by putting the index of the pixel (in the crossdisp axis) that fails the test in the message and change from "beyond" to "above" or "below" (although this might be a bit confusing if the axes are flipped).  This then means that the warning string is likely to be unique as the user tries to make adjustments and so will repeatedly be shown automatically.  The warning now looks something like:

```
background region extends above image boundaries (44.04264382881123 >= 44)
```

Reply:
I think sticking with "beyond" will be more future-proof, especially since the parenthetical makes it apparent which side of the image is being referenced.

Do you think it's helpful to clip the number of decimal points (something like `{windows_max:.02f}`)? Leaving the whole thing seems like too much, but the more you clip, the higher the chance the next message will not be unique.

This version of the message still stops at stating the issue. Do you remain opposed to giving advice on how to fix the warning?

Reply:
> This version of the message still stops at stating the issue. Do you remain opposed to giving advice on how to fix the warning?

I tried to address this concern with changing to above/below (since for the two-sided case, that provides some more information that hints at which side is to "blam"), but I'm happy to revert that.  

I don't think we can directly provide any specific advice since there are many approaches the user might want to take.  We could say something super generic, like "consider revising the input trace(s) and/or separations and/or widths if you want to ensure the entire background window is within the image", but does that really add enough information/context to be worth the longer message?  

If we're lengthening the message, I think it would be more useful to add logic that computes the minimum percentage of the entire window within the image, or something like that... but I think that can be deferred until later.  Another option would be to suggest plotting the window or weights to inspect, but there currently isn't built-in functionality to do that.

> Do you think it's helpful to clip the number of decimal points (something like {windows_max:.02f})?  Leaving the whole thing seems like too much, but the more you clip, the higher the chance the next message will not be unique.

I considered that but had the same thought - the extra length increases the chances of a unique message and truncating could result in `-0.00 < 0`, which would be confusing.